src/style/style_layer/heatmap_style_layer.js

```diff
@@ -0,0 +1,58 @@
+// @flow
+
+const StyleLayer = require('../style_layer');
+const HeatmapBucket = require('../../data/bucket/heatmap_bucket');
+const RGBAImage = require('../../util/image').RGBAImage;
+
+import type Texture from '../../render/texture';
+
+class HeatmapStyleLayer extends StyleLayer {
+
+    heatmapTexture: ?WebGLTexture;
+    heatmapFbo: ?WebGLFramebuffer;
+    colorRampData: Uint8Array;
+    colorRamp: RGBAImage;
+    colorRampTexture: ?Texture;
+
+    createBucket(options: any) {
+        return new HeatmapBucket(options);
+    }
+
+    constructor(layer: LayerSpecification) {
+        super(layer);
+        this.colorRampData = new Uint8Array(256 * 4);
+        if (!this.getPaintProperty('heatmap-color')) {
+            this.setPaintProperty('heatmap-color', this._paintSpecifications['heatmap-color'].default, '');
+        }
+    }
+
+    _applyPaintDeclaration(name: any, declaration: any, options: any, globalOptions: any, animationLoop: any, zoomHistory: any) {
+        super._applyPaintDeclaration(name, declaration, options, globalOptions, animationLoop, zoomHistory);
+        if (name === 'heatmap-color') {
+            const len = this.colorRampData.length;
```

Comment:
I actually started with `setPaintProperty` but later moved to this, I don't remember why — will recheck.

Reply:
Oh, I think I remember now — the problem is you can't call `getPaintValue` immediately after `setPaintProperty`, because it's being set up in a separate "apply paint transitions" stage.

Reply:
Ah, okay.  Could we add a comment about that here, so that no one is tempted to try such a refactor later on?